lib/ansible/module_utils/aws/urls.py

```diff
@@ -0,0 +1,166 @@
+# (C) Unknown?
+
+import datetime
+import hashlib
+import hmac
+import operator
+
+from ansible.module_utils.urls import open_url
+from ansible.module_utils.ec2 import get_aws_connection_info
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+
+
+def hexdigest(s):
+    """
+    Returns the sha256 hexdigest of a string after encoding.
+    """
+    return hashlib.sha256(s.encode('utf-8')).hexdigest()
+
+
+def format_querystring(params=None):
+    """
+    Returns properly url-encoded query string from the provided params dict.
+
+    It's specially sorted for cannonical requests
+    """
+
+    if not params:
+        return ""
+
+    # Query string values must be URL-encoded (space=%20). The parameters must be sorted by name.
+    return urlencode(sorted(params.items(), operator.itemgetter(0)))
+
+
+# Key derivation functions. See:
+# http://docs.aws.amazon.com/general/latest/gr/signature-v4-examples.html#signature-v4-examples-python
+def sign(key, msg):
+    '''
+    Return digest for key applied to msg
+    '''
+    return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()
+
+
+def get_signature_key(key, dateStamp, regionName, serviceName):
+    '''
+    Returns signature key for AWS resource
+    '''
+    kDate = sign(('AWS4' + key).encode('utf-8'), dateStamp)
+    kRegion = sign(kDate, regionName)
+    kService = sign(kRegion, serviceName)
+    kSigning = sign(kService, 'aws4_request')
+    return kSigning
+
+
+# Reference: https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html
+def signed_request(method="GET", service=None, host=None, uri=None, query=None, body="", module=None, boto3=True, headers=None):
+    """Generate a SigV4 request to an AWS resource for a module
+
+    This is used if you wish to authenticate with AWS credentials to a secure endpoint like an elastisearch domain.
+
+    Returns :class:`HTTPResponse` object.
+
+    Example:
+        result = signed_request(
+            module=this,
+            service="es",
+            host="search-recipes1-xxxxxxxxx.us-west-2.es.amazonaws.com",
+        )
+
+    :kwarg host: endpoint to talk to
+    :kwarg service: AWS id of service (like `ec2` or `es`)
+    :kwarg module: An AnsibleAWSModule to gather connection info from
+
+    :kwarg body: (optional) Payload to send
+    :kwarg method: (optional) HTTP verb to use
+    :kwarg query: (optional) dict of query params to handle
+    :kwarg uri: (optional) Resource path without query parameters
+    :returns: HTTPResponse
+    """
+
+    # "Constants"
+
+    t = datetime.datetime.utcnow()
+    amz_date = t.strftime('%Y%m%dT%H%M%SZ')
+    datestamp = t.strftime('%Y%m%d')  # Date w/o time, used in credential scope
+    algorithm = 'AWS4-HMAC-SHA256'
+
+    # AWS stuff
+
+    region, dummy, boto_params = get_aws_connection_info(module, boto3)
+
+    access_key = boto_params["aws_access_key_id"]
```

Comment:
I don't think boto_params will include access_key or secret_key if a profile is being used (which might be through passing `profile` parameter or with `AWS_PROFILE` set)

Looking at https://github.com/jmenga/requests-aws-sign the way to do it is to use 
```
session = session.Session()
credentials = session.get_credentials()
```
but even there, we'd need to cope with the `profile` parameter.

Reply:
I'm not sure how to deal with the profile parameter. Should this be handled within `get_aws_connection_info`?

Reply:
`get_aws_connection_info` will return you a `profile` inside `boto_params` and normally `boto3_conn` would then take care of the rest - you can see how `profile` is handled at https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L122-L139

With that `session`, you can then use `get_credentials` as above.